lib/devise-security/models/secure_validatable.rb

```diff
@@ -5,73 +5,114 @@
 
 module Devise
   module Models
-    # SecureValidatable creates better validations with more validation for security
+    # SecureValidatable adds additional validations for better security.
+    # @note depends on database_authenticatable
+    #
+    # == Validations
+    # === Unique login
+    # Verifies that the first +authentication_key+ is unique (scoped by any
+    # additional keys). This respects the +Devise.case_sensitive_keys+
+    # configuration.  This validation is skipped if the model already declares a
+    # +validates_uniqueness_of+ validator on the first +authentication_key+.
+    # @note You should ensure that there is a unique index on +login_attribute+
+    #   to avoid race conditions.
+    #
+    # === Email validation
+    # If the model requires an email address (i.e, +email_required?+ return +true+),
+    # then the email address is verified to be present and unique.
+    #
+    # === Password Complexity
+    # Enforces password complexity requirements as described in {PasswordComplexityValidator}.
+    #
+    # === Password re-use
+    # Disallows re-using current password when changing the password.  This
+    # validation is disabled if the {PasswordArchivable} module is used because
+    # it results in redundant validation failures.
     #
     # == Options
     #
-    # SecureValidatable adds the following options to devise_for:
+    # SecureValidatable adds the following options to the +devise+ declaration:
     #
-    #   * +email_regexp+: the regular expression used to validate e-mails;
-    #   * +password_length+: a range expressing password length. Defaults from devise
-    #   * +password_regex+: need strong password. Defaults to /(?=.*\d)(?=.*[a-z])(?=.*[A-Z])/
+    # * +password_complexity+: descriptor of the minimum character classes
+    #   required for password complexity, i.e.:
+    #    { digits: 1, lower: 1, upper: 1, symbol 1}
+    # * +email_validation+: the validation options to be passed to the email validator
+    # * +password_length+: a +Range+ expressing password length. Defaults from devise
     #
     module SecureValidatable
+      extend ActiveSupport::Concern
       include Devise::Models::Compatibility
 
-      def self.included(base)
-        base.extend ClassMethods
+      included do |base|
         assert_secure_validations_api!(base)
 
-        base.class_eval do
-          already_validated_email = false
+        devise :database_authenticatable
 
-          # validate login in a strict way if not yet validated
-          unless has_uniqueness_validation_of_login?
-            validation_condition = "#{login_attribute}_changed?".to_sym
+        already_validated_email = false
 
+        # login attribute should be unique
+        unless has_uniqueness_validation_of_login?
+          # Only perform uniqueness check when the login attribute is dirty so
+          # we can avoid hitting the database unnecessarily.
+          if defined?(will_save_change_to_attribute?)
```

Comment:
This was a pretty hacky-fix to get a test passing. Not sure if we should use an older form of dirty checking in the `else` block.

Reply:
This is also needed for Rails updates IIRC as they changed the method name between recent versions.

Reply:
Yup looks like my memory is correct, see line 97 and line 98.